akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -444,108 +568,173 @@ private[akka] class Shard(
-        val (id, _) = extractEntityId(msg)
-        if (entities.entityIdExists(id)) {
-          if (VerboseDebug) log.debug("Entity already known about. Try and deliver. [{}]", id)
-          deliverMessage(msg, sender(), OptionVal.Some(entityIds))
-        } else {
-          if (VerboseDebug) log.debug("New entity, add it to batch of pending starts. [{}]", id)
-          appendToMessageBuffer(id, msg, sender())
-          context.become(waitingForUpdate(pendingStarts + (id -> None)))
-        }
-      case msg =>
-        // shouldn't be any other message types, but just in case
-        log.warning(
-          "Stashing unexpected message [{}] while waiting for remember entities update of [{}]",
-          msg.getClass,
-          entityIds.mkString(", "))
-        stash()
+          log.debug("Start entity [{}] while a write already in progress. Marking as pending", entityId)
+        entities.rememberingStart(entityId, ackTo = Some(sender))
+      } else {
+        // it's already running, ack immediately
+        sender() ! ShardRegion.StartEntityAck(entityId, shardId)
+      }
+
+    // below cases should handle same messages as in idle
+    case Terminated(ref) =>
+      entities.entityId(ref) match {
+        case OptionVal.Some(entityId) =>
+          entities.entityState(entityId) match {
+            case RememberingStop(_) =>
+              log.debug("Stop of [{}] arrived, already is among the pending stops.", entityId)
+            case Active(_) =>
+              // FIXME should trigger backoff + restart?
+              log.warning("Stop of [{}] arrived, should lead to backoff restart, not implemented yet.", entityId)
+            case Passivating(_) =>
+              // will go in next batch update
+              log.debug("Stop of [{}] arrived while updating, adding it to batch of pending stops.", entityId)
+              entities.rememberingStop(entityId, Passivating)
+            case unexpected =>
+              log.warning("Got a terminated for [{}], entityId [{}] which is in state [{}]", ref, entityId, unexpected)
+          }
+        case OptionVal.None =>
+          // FIXME what would this Terminated be? Remember entities store stop?
+          stash()
+      }
+    case _: CoordinatorMessage         => stash()
+    case _: StartEntityInternal        => stash()
+    case _: RememberEntityCommand      => stash()
+    case l: LeaseLost                  => receiveLeaseLost(l)
+    case _: ShardRegion.StartEntityAck =>
+      // FIXME if we got an ack it was started elsewhere we should probably make sure we don't have it in current or pending starts?
+      stash()
+    case ShardRegion.Passivate(stopMessage) =>
+      if (VerboseDebug)
+        log.debug(
+          "Passivation of [{}] arrived while updating.",
+          entities.entityId(sender()).getOrElse(s"Unknown actor ${sender()}"))
+      passivate(sender(), stopMessage)
+    case _: ShardRegionCommand           => stash()
```

Comment:
We don't ever get this do we? The only shardrefioncommand we do is passivated anddd that is handled above

Reply:
Good point. I'll remove it.